### ISSUE TITLE
[core] Remove deprecated APIs for 6.0.0

### ISSUE THREAD

**Issue by jsotuyod** [pmd/pmd#463]
List of deprecated APIs to be removed:

* [x] `Renderer.getPropertyDefinitions`
* [x] `AbstractRenderer.defineProperty(String, String)`
* [x] `AbstractRenderer.propertyDefinitions`
* [x] `ReportListener`
* [x] `Report.addListener(ReportListener)`
* [x] `SynchronizedReportListener`
* [x] `CPDConfiguration.CPDConfiguration(int, Language, String)`
* [x] `CPDConfiguration.getRendererFromString(String)`
* [x] `StreamUtil`
* [x] `StringUtil.appendXmlEscaped(StringBuilder, String)`
* [x] `StringUtil.htmlEncode(String)`

**Comment by oowekyala** [comment]
I'd like to mention these as candidates to deprecation, as they will not be used anymore either in pmd and the eclipse plugin when #479 and pmd/pmd-eclipse-plugin#25 are merged:
- [ ] `CollectionUtil.addWithoutDuplicates(T[], T)`
- [ ] `CollectionUtil.addWithoutDuplicates(T[], T[])`
- [ ] `CollectionUtil.areSemanticEquals(T[], T[])`

Furthermore:
- [ ] `CollectionUtil.areEqual(Object, Object)`
- [ ] `CollectionUtil.arraysAreEqual(Object, Object)`
- [ ] `CollectionUtil.valuesAreTransitivelyEqual(Object[], Object[])`

The first method is used only once in PMD. Together, the three call each other in a recursive cycle. They're used once in the eclipse plugin (in `RuleUtil`). I think we should consider replacing the only usage of `areEqual` with `Object.equals` in PMD, and move these methods to the plugin's own utility classes, which are to be created. 

Maybe we could even deprecate them for 5.8.1 and remove them in 6.0.0?

Last but not least, the entire class `net.sourceforge.pmd.lang.rule.properties.PropertyDescriptorFactory` is not used either in PMD or the plugin. I think its functionality was superseded by `PropertyDescriptorUtil` at some point anyway. I think it should be removed too.

**Comment by jsotuyod** [comment]
@oowekyala very interesting analysis! Thanks for the input!

I believe we should deprecate those before removing in a following major release (7.0.0). People may be using those in ad-hoc proprietary plugins with custom rules / languages.

I don't think deprecating those in 5.8.1 would be wise. Firstly because it's a bug fix release. Secondly, because it's probably the last release before 6.0.0, beating the point of having a fair warning time before breaking the API.

**Comment by oowekyala** [comment]
@jsotuyod Right that totally makes sense!

**Comment by jsotuyod** [comment]
[This method](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/PMD.java#L352-L375) is not actually deprecated, but we should definitely get rid of it.

**Comment by oowekyala** [comment]
To be honest all of `StringUtil` could be deprecated in favour of `lang3.StringUtils`. Many methods here pretend to be faster than standard library but were probably written 10 years ago (some are said to be dated from java 1.3), and seeing as they're used a few times max we should not bother making our own library for that.
